### ISSUE TITLE
Recent ndk-sys yank is breaking CI.

### ISSUE THREAD

**Issue by Zageron** [rust-mobile/ndk#369]
Checksum is failing on ndk-sys 0.4.0.

https://github.com/Zageron/rrr/actions/runs/3529598287/jobs/5920752227

Etc...

**Comment by ogoffart** [comment]
So it seems that the problem is that there is a version `0.4.0` and a version `0.4.0+25.0.8775105` on crates.io.
Apparently, the code in cargo get confused because the two version compare the same. 

 Cf  https://github.com/rust-lang/cargo/issues/6504  https://github.com/rust-lang/crates.io/issues/1059 https://github.com/rust-lang/cargo/issues/7180

The immediate solution would be to release ndk-sys 0.4.1 as soon as possible (I recommend re-publishing 0.4.0 as 0.4.1)

**Comment by MarijnS95** [comment]
Duplicate of #368, and the offending crate has been yanked: I cannot repro this locally with a `cargo update` on a crate that consumes `ndk-sys`.

Could this possibly exhibit itself when an a `cargo update` (or in general downloading the initial crates.io index) was performed _while the broken `0.4.0+25.0.8775105` release was live_, leading to the crate being downloaded into `~/.cargo` and considered by commands [like `cargo metadata`](https://github.com/Zageron/rrr/actions/runs/3529598287/jobs/5920752227#step:8:180) going forward? That sounds like a followup `cargo` bug if it isn't able to recover from having previously downloaded a broken crate that's now been yanked, when it should instead be disregarded completely.
(None of the links shared by @ogoffart seem to describe this, but crates.io should disallow such publishes anyway :no_mouth:)

I was anyway planning on releasing a `0.4.1+23.1.7779620` to represent the version of the previous release, if that helps. Going forward we'll have to think more carefully about bumping this metadata version as our CI will immediately publish it as a release; any version upgrades should just be a release at that point.

**Comment by ogoffart** [comment]
I think there is a race condition in cargo and the bug happens only sometimes when downloading the crates from fresh cargo cache.
But now that 0.4.1+23.1.7779620 is released, the problem should be solved.
thank you!

**Comment by MarijnS95** [comment]
> I think there is a race condition in cargo

Is this reported upstream somewhere?

> and the bug happens only sometimes when downloading the crates from fresh cargo cache.

Right, the CI run linked by @Zageron above indicates [`No cache found.`](https://github.com/Zageron/rrr/actions/runs/3529598287/jobs/5920752227#step:7:57); that CI run is from a few hours ago, well after I yanked the broken release yesterday which is strange to say the least.

> But now that 0.4.1+23.1.7779620 is released, the problem should be solved. thank you!

Thanks, I'll close this as completed then unless you want to confirm it first!
(Confirmed to build and run fine on my crates)

**Comment by MarijnS95** [comment]
Hah I'm also running into the exact same issue randomly, perhaps because of still being `Cargo.lock`'ed into `0.4.0`, despite (again) this crate being yanked well over a day ago and the ~/.cargo` cache being completely empty/fresh. :disappointed:

**Comment by MarijnS95** [comment]
Reopening this because it's super easy to repro. On a cleaned `~/.cargo`, all I need to do is force use of `0.4.0` (not `0.4.1`!) with the semver string `"=0.4.0"` or use a `Cargo.lock` fixed (not yet updated) to `0.4.1`. Then try to build, and before it even downloads the crate:

```
  Downloaded ndk-sys v0.4.0
error: failed to verify the checksum of `ndk-sys v0.4.0`
```

(There's no `ndk-sys` folder _anywhere_ inside `~/.cargo/registry/{cache,src}/github.com-1ecc6299db9ec823/`)

**Comment by eminence** [comment]
See also: https://github.com/rust-lang/cargo/issues/11412

**Comment by brianm-sra** [comment]
Please update file ndk-glue/Cargo.toml to reference version ndk-sys v0.4.1 

On https://github.com/rust-lang/cargo/issues/7180 it is mentioned

"I have since published 0.4.1 to help users resolve this issue,"

**Comment by MarijnS95** [comment]
Not hearing anyone about this issue anymore, hoping it is resolved now (and will be much more careful pushing version changes in the future)! :crossed_fingers:
